File: algorithms/[77]组合/solution.py
# topics = ["回溯算法"]

from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res: List[List[int]] = []
        path: List[int] = []

        def backtrack(idx: int = 1) -> None:
            size = len(path)
            if size == k:
                res.append(path[:])
                return
            for i in range(idx, n + 1):
                if size > k:
                    return
                path.append(i)
                backtrack(i + 1)
                path.pop()

        backtrack()

        return res

File: algorithms/[77]组合/solution.ts
// topics = ["回溯算法"]

function combine(n: number, k: number): number[][] {
  const res: number[][] = [];
  const path: number[] = [];

  const backtrack = (idx = 1): void => {
    const len = path.length;
    if (len == k) {
      res.push([...path]);
      return;
    }
    for (let i = idx; i <= n; i += 1) {
      if (len > k) {
        return;
      }
      path.push(i);
      backtrack(i + 1);
      path.pop();
    }
  };

  backtrack();

  return res;
}
